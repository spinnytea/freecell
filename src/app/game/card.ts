// XXX (techdebt) duplicate
const BOTTOM_OF_CASCADE = 99;

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];
export const isAdjacent = ({ min, max }: { min: Rank; max: Rank }) =>
	RankList.indexOf(min) === RankList.indexOf(max) - 1;

// REVIEW (techdebt) is "fixture" the right name?
//  - cell -> freecell
//  - foundation -> homecell
//  - cascade -> column
export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

export type MoveSourceType = 'deck' | 'cell' | 'foundation' | 'cascade:single' | 'cascade:sequence';
export type MoveDestinationType = 'cell' | 'foundation' | 'cascade:empty' | 'cascade:sequence';
export const MoveDestinationTypePriorities: {
	[moveSourceType in MoveSourceType]: { [moveDestinationType in MoveDestinationType]: number };
} = {
	// XXX (controls) deck: when will we get to do this?
	'deck': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cell': {
		'cell': 1,
		'foundation': 3,
		'cascade:empty': 2,
		'cascade:sequence': 4,
	},
	// XXX (controls) foundation: down from foundation means "back into play?"
	'foundation': {
		'cell': 1,
		'foundation': 4,
		'cascade:empty': 2,
		'cascade:sequence': 3,
	},
	'cascade:single': {
		'cell': 2,
		'foundation': 3,
		'cascade:empty': 1,
		'cascade:sequence': 4,
	},
	'cascade:sequence': {
		'cell': 1,
		'foundation': 2,
		'cascade:empty': 3,
		'cascade:sequence': 4,
	},
};

/**
	REVIEW (techdebt) does AvailableMove need to store moveDestinationType and priority?
	 - moveDestinationType is needed for building the list (like moveSourceType)
	   but do we need to store it?
	 - priority could be build into the sort order of the list (pick the first item),
	   rather than stored with the location (pick the highest number)
	 - it is nice to see in the DebugCursors the different categories
	 	priority: 'high' | 'low';
	---
	 - we have sufficient impl to simplify now
	 - wait for unit testing of DebugCursors
	 - wait for manualtesting of DebugCursors
*/
export interface AvailableMove {
	/** where we could move the card */
	location: CardLocation;
	/** helps us think about priorities / communicate settings */
	moveDestinationType: MoveDestinationType;
	/** if we are going to visualize them debug mode, we need to have it precomputed */
	priority: number;
}

/**
	foundation: h
	cells: a - d
	cascades: 1 - 9, t (1-8, but we allow 9 and 10 columns)

	@see [Standard FreeCell Notation](https://www.solitairelaboratory.com/solutioncatalog.html)
*/
export type Position =
	| 'a'
	| 'b'
	| 'c'
	| 'd'
	| 'h'
	| '1'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| 't';

export interface Card {
	rank: Rank;
	suit: Suit;
	location: CardLocation;
}

export interface CardSequence {
	/**
		the first card in the sequence
		(cursor, selection)
	*/
	location: CardLocation;

	/**
		the list of cards in the sequence
		(alternates red/black, rank descends by one)
	*/
	cards: Card[];

	/** since we are allowing any card to be selected or inspected, not all of them are movable */
	canMove: boolean;
}

export function isLocationEqual(a: CardLocation, b: CardLocation) {
	return a.fixture === b.fixture && a.data[0] === b.data[0] && a.data[1] === b.data[1];
}

export function shorthandCard(card: Card | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank === 'joker' ? 'W' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}

export function parseShorthandCard(
	r: string | undefined,
	s: string | undefined
): { rank: Rank; suit: Suit } | null {
	if (r === ' ' && s === ' ') return null;

	let rank: Rank;
	let suit: Suit;
	switch (r) {
		case 'A':
			rank = 'ace';
			break;
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			rank = r;
			break;
		case 'T':
			rank = '10';
			break;
		case 'J':
			rank = 'jack';
			break;
		case 'Q':
			rank = 'queen';
			break;
		case 'K':
			rank = 'king';
			break;
		case 'W':
			rank = 'joker';
			break;
		default:
			throw new Error(`invalid rank shorthand: "${r ?? 'undefined'}"`);
	}

	switch (s) {
		case 'C':
			suit = 'clubs';
			break;
		case 'D':
			suit = 'diamonds';
			break;
		case 'H':
			suit = 'hearts';
			break;
		case 'S':
			suit = 'spades';
			break;
		default:
			throw new Error(`invalid suit shorthand: "${s ?? 'undefined'}"`);
	}

	return { rank, suit };
}

export function shorthandSequence(sequence: CardSequence, includePosition = false) {
	const cards = sequence.cards.map((card) => shorthandCard(card)).join('-');

	if (sequence.canMove && includePosition) {
		return shorthandPosition(sequence.location) + ' ' + cards;
	}

	return cards;
}

export function shorthandPosition(location: CardLocation): Position {
	const d0 = location.data[0];
	if (location.fixture === 'foundation') {
		return 'h';
	} else if (location.fixture === 'cascade') {
		// this doesn't check data[1], it assumes it's the final row
		// sequences would need to check data[1] + card.length, not just the location
		// (could pass in a optional canMove with default of true)
		if (d0 === 9) {
			return 't';
		} else if (d0 >= 0 && d0 < 9) {
			return (d0 + 1).toString(10) as Position;
		}
	} else if (location.fixture === 'cell') {
		if (d0 >= 0 && d0 < 6) {
			return (d0 + 10).toString(16) as Position;
		}
	}
	throw new Error(`invalid position: ${JSON.stringify(location)}`);
}

/**
	this is part 1 of a 2 step process
	moves are not always obvious
	i.e. 1h - _which_ foundation do we use? there are 4
	i.e. 42 - is this a cascade:single or cascade:sequence moving to a cascade:sequence or cascade:empty?

	notice also that this function only accepts the single character, it does not accept a game
	so which d1 do we use for a cascade? this will return an invalid value (too high), which will be clamped if used directly
*/
export function parseShorthandPosition_INCOMPLETE(p: string | undefined): CardLocation {
	switch (p) {
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			// this isn't a valid cursor position, it will need to be clamped
			// cascades can have sequences, so you need to decide if you really want the "bottom"
			return { fixture: 'cascade', data: [parseInt(p, 10) - 1, BOTTOM_OF_CASCADE] };
		case 't':
			return { fixture: 'cascade', data: [9, BOTTOM_OF_CASCADE] };
		case 'h':
			// h could refer to _any_ of the foundations; this needs to be verified
			return { fixture: 'foundation', data: [0] };
		case 'a':
		case 'b':
		case 'c':
		case 'd':
		case 'e':
		case 'f':
			return { fixture: 'cell', data: [parseInt(p, 16) - 10] };
		default:
			throw new Error(`invalid position shorthand: "${p ?? 'undefined'}"`);
	}
}
