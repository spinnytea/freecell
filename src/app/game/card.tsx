export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades';
export const SuitList: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades'];
export const isRed = (suit: Suit) => suit === 'diamonds' || suit === 'hearts';

export type Rank =
	| 'ace'
	| '2'
	| '3'
	| '4'
	| '5'
	| '6'
	| '7'
	| '8'
	| '9'
	| '10'
	| 'jack'
	| 'queen'
	| 'king'
	| 'joker';
export const RankList: Rank[] = [
	'ace',
	'2',
	'3',
	'4',
	'5',
	'6',
	'7',
	'8',
	'9',
	'10',
	'jack',
	'queen',
	'king',
];

export type Fixture = 'deck' | 'cell' | 'foundation' | 'cascade';
export interface CardLocation {
	fixture: Fixture;
	data: number[];
}

export interface Card {
	rank: Rank;
	suit: Suit;
	location: CardLocation;
}

export function shorthand(card: Card | null | undefined) {
	if (!card) return '  ';
	const r = card.rank === '10' ? 'T' : card.rank[0];
	const s = card.suit[0];
	return (r + s).toUpperCase();
}
