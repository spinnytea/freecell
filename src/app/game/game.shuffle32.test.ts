import { shorthandCard } from '@/app/game/card/card';
import { FreeCell } from '@/app/game/game';

describe('game.shuffle32', () => {
	test('Game #1', () => {
		let game = new FreeCell();
		expect(game.deck.length).toBe(52);
		game = game.shuffle32(1);
		expect(game.deck.length).toBe(52);
		expect(game.previousAction).toEqual({ text: 'shuffle deck (1)', type: 'shuffle' });
		expect(shorthandCard(game.deck[51])).toBe('JD');
		expect(shorthandCard(game.deck[50])).toBe('2D');
		expect(shorthandCard(game.deck[49])).toBe('9H');
		expect(shorthandCard(game.deck[48])).toBe('JC');
		expect(shorthandCard(game.deck[3])).toBe('6S');
		expect(shorthandCard(game.deck[2])).toBe('9C');
		expect(shorthandCard(game.deck[1])).toBe('2H');
		expect(shorthandCard(game.deck[0])).toBe('6H');
		expect(game.deck[51]).toEqual({
			rank: 'jack',
			suit: 'diamonds',
			location: { fixture: 'deck', data: [51] },
		});
		expect(game.deck[0]).toEqual({
			rank: '6',
			suit: 'hearts',
			location: { fixture: 'deck', data: [0] },
		});
		game = game.dealAll();
		expect(game.previousAction).toEqual({ text: 'deal all cards', type: 'deal' });
		expect(game.deck.length).toBe(0);
		expect(game).toMatchSnapshot();
		expect(game.tableau[0].length).toBe(7);
		expect(game.tableau[1].length).toBe(7);
		expect(game.tableau[2].length).toBe(7);
		expect(game.tableau[3].length).toBe(7);
		expect(game.tableau[4].length).toBe(6);
		expect(game.tableau[5].length).toBe(6);
		expect(game.tableau[6].length).toBe(6);
		expect(game.tableau[7].length).toBe(6);
		expect(shorthandCard(game.tableau[0][0])).toBe('JD');
		expect(shorthandCard(game.tableau[1][0])).toBe('2D');
		expect(shorthandCard(game.tableau[2][0])).toBe('9H');
		expect(shorthandCard(game.tableau[3][0])).toBe('JC');
		expect(shorthandCard(game.tableau[0][6])).toBe('6S');
		expect(shorthandCard(game.tableau[1][6])).toBe('9C');
		expect(shorthandCard(game.tableau[2][6])).toBe('2H');
		expect(shorthandCard(game.tableau[3][6])).toBe('6H');
		expect(game.tableau[0][0]).toEqual({
			rank: 'jack',
			suit: 'diamonds',
			location: { fixture: 'cascade', data: [0, 0] },
		});
		expect(game.tableau[3][6]).toEqual({
			rank: '6',
			suit: 'hearts',
			location: { fixture: 'cascade', data: [3, 6] },
		});
		expect(game.print()).toBe(
			'' +
				'>                        \n' +
				' JD 2D 9H JC 5D 7H 7C 5H \n' +
				' KD KC 9S 5S AD QC KH 3H \n' +
				' 2S KS 9D QD JS AS AH 3C \n' +
				' 4C 5C TS QH 4H AC 4D 7S \n' +
				' 3S TD 4S TH 8H 2C JH 7D \n' +
				' 6D 8S 8D QS 6C 3D 8C TC \n' +
				' 6S 9C 2H 6H             \n' +
				' deal all cards'
		);
	});

	test('Game #617', () => {
		let game = new FreeCell();
		game = game.shuffle32(617);
		expect(game.previousAction).toEqual({ text: 'shuffle deck (617)', type: 'shuffle' });
		game = game.dealAll();
		expect(game.previousAction).toEqual({ text: 'deal all cards', type: 'deal' });
		expect(game.print()).toBe(
			'' +
				'>                        \n' +
				' 7D AD 5C 3S 5S 8C 2D AH \n' +
				' TD 7S QD AC 6D 8H AS KH \n' +
				' TH QC 3H 9D 6S 8D 3D TC \n' +
				' KD 5H 9S 3C 8S 7H 4D JS \n' +
				' 4C QS 9C 9H 7C 6H 2C 2S \n' +
				' 4S TS 2H 5D JC 6C JH QH \n' +
				' JD KS KC 4H             \n' +
				' deal all cards'
		);
	});

	test('Game #11982', () => {
		const game = new FreeCell().shuffle32(11982);
		const matchSeed = /shuffle deck \((\d+)\)/.exec(game.previousAction.text);
		expect(matchSeed).not.toBe(null);
		expect(matchSeed?.[1]).toMatch(/^\d+$/);
		expect(matchSeed?.[1]).not.toBe('11982');
	});

	test('partial deck', () => {
		const game = FreeCell.parse(
			'' + //
				'>JC          9C 9D 9H 9S \n' + //
				'                         \n' + //
				' hand-jammed'
		);

		expect(game.print()).toBe(
			'' + //
				'>JC          9C 9D 9H 9S \n' + //
				'                         \n' + //
				':d KS KH KD KC QS QH QD QC JS JH JD TS TH TD TC \n' + //
				' hand-jammed'
		);
		expect(game.shuffle32(0).print()).toBe(
			'' + //
				'>JC          9C 9D 9H 9S \n' + //
				'                         \n' + //
				':d QD JH QH TD TC QC JD KH JS QS TS KD TH KC KS \n' + //
				' shuffle deck (0)'
		);
		expect(game.shuffle32(0).shuffle32(0).print()).toBe(
			'' + //
				'>JC          9C 9D 9H 9S \n' + //
				'                         \n' + //
				':d JD QS QC KC KS KH TS JH JS TC KD QH TH TD QD \n' + //
				' shuffle deck (0)'
		);
	});
});
