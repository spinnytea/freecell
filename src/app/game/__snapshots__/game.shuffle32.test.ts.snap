// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game.shuffle32 Game #1 1`] = `
FreeCell {
  "availableMoves": null,
  "cards": [
    {
      "location": {
        "data": [
          5,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "ace",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          4,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "ace",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          6,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "ace",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          5,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "ace",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          5,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "2",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          1,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "2",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          2,
          6,
        ],
        "fixture": "cascade",
      },
      "rank": "2",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          0,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "2",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          7,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "3",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          5,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "3",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          7,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "3",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          0,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "3",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          0,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "4",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          6,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "4",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          4,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "4",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          2,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "4",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          1,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "5",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          4,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "5",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          7,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "5",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          3,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "5",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          4,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "6",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          0,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "6",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          3,
          6,
        ],
        "fixture": "cascade",
      },
      "rank": "6",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          0,
          6,
        ],
        "fixture": "cascade",
      },
      "rank": "6",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          6,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "7",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          7,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "7",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          5,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "7",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          7,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "7",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          6,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "8",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          2,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "8",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          4,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "8",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          1,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "8",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          1,
          6,
        ],
        "fixture": "cascade",
      },
      "rank": "9",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          2,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "9",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          2,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "9",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          2,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "9",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          7,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "10",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          1,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "10",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          3,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "10",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          2,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "10",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          3,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "jack",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          0,
          0,
        ],
        "fixture": "cascade",
      },
      "rank": "jack",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          6,
          4,
        ],
        "fixture": "cascade",
      },
      "rank": "jack",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          4,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "jack",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          5,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "queen",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          3,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "queen",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          3,
          3,
        ],
        "fixture": "cascade",
      },
      "rank": "queen",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          3,
          5,
        ],
        "fixture": "cascade",
      },
      "rank": "queen",
      "suit": "spades",
    },
    {
      "location": {
        "data": [
          1,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "king",
      "suit": "clubs",
    },
    {
      "location": {
        "data": [
          0,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "king",
      "suit": "diamonds",
    },
    {
      "location": {
        "data": [
          6,
          1,
        ],
        "fixture": "cascade",
      },
      "rank": "king",
      "suit": "hearts",
    },
    {
      "location": {
        "data": [
          1,
          2,
        ],
        "fixture": "cascade",
      },
      "rank": "king",
      "suit": "spades",
    },
  ],
  "cells": [
    null,
    null,
    null,
    null,
  ],
  "cursor": {
    "data": [
      0,
    ],
    "fixture": "cell",
  },
  "deck": [],
  "foundations": [
    null,
    null,
    null,
    null,
  ],
  "previousAction": {
    "text": "deal all cards",
    "type": "deal",
  },
  "selection": null,
  "tableau": [
    [
      {
        "location": {
          "data": [
            0,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "jack",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            0,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "king",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            0,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "2",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            0,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "4",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            0,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "3",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            0,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "6",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            0,
            6,
          ],
          "fixture": "cascade",
        },
        "rank": "6",
        "suit": "spades",
      },
    ],
    [
      {
        "location": {
          "data": [
            1,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "2",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            1,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "king",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            1,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "king",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            1,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "5",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            1,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "10",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            1,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "8",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            1,
            6,
          ],
          "fixture": "cascade",
        },
        "rank": "9",
        "suit": "clubs",
      },
    ],
    [
      {
        "location": {
          "data": [
            2,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "9",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            2,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "9",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            2,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "9",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            2,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "10",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            2,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "4",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            2,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "8",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            2,
            6,
          ],
          "fixture": "cascade",
        },
        "rank": "2",
        "suit": "hearts",
      },
    ],
    [
      {
        "location": {
          "data": [
            3,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "jack",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            3,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "5",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            3,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "queen",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            3,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "queen",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            3,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "10",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            3,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "queen",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            3,
            6,
          ],
          "fixture": "cascade",
        },
        "rank": "6",
        "suit": "hearts",
      },
    ],
    [
      {
        "location": {
          "data": [
            4,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "5",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            4,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "ace",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            4,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "jack",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            4,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "4",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            4,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "8",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            4,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "6",
        "suit": "clubs",
      },
    ],
    [
      {
        "location": {
          "data": [
            5,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "7",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            5,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "queen",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            5,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "ace",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            5,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "ace",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            5,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "2",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            5,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "3",
        "suit": "diamonds",
      },
    ],
    [
      {
        "location": {
          "data": [
            6,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "7",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            6,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "king",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            6,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "ace",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            6,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "4",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            6,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "jack",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            6,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "8",
        "suit": "clubs",
      },
    ],
    [
      {
        "location": {
          "data": [
            7,
            0,
          ],
          "fixture": "cascade",
        },
        "rank": "5",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            7,
            1,
          ],
          "fixture": "cascade",
        },
        "rank": "3",
        "suit": "hearts",
      },
      {
        "location": {
          "data": [
            7,
            2,
          ],
          "fixture": "cascade",
        },
        "rank": "3",
        "suit": "clubs",
      },
      {
        "location": {
          "data": [
            7,
            3,
          ],
          "fixture": "cascade",
        },
        "rank": "7",
        "suit": "spades",
      },
      {
        "location": {
          "data": [
            7,
            4,
          ],
          "fixture": "cascade",
        },
        "rank": "7",
        "suit": "diamonds",
      },
      {
        "location": {
          "data": [
            7,
            5,
          ],
          "fixture": "cascade",
        },
        "rank": "10",
        "suit": "clubs",
      },
    ],
  ],
  "win": false,
}
`;
